csrc/multidevice/utils.cpp

```diff
@@ -254,7 +217,13 @@ void insertReshardings(Fusion* fusion) {
   // Remove this after we refactor this as a pre-segmenter pass.
   FusionGuard fg(fusion);
   auto exprs = fusion->exprs();
+  // Replacing expression inputs creates a new expression. Map the old
+  // expression to new expression, so that we can continue iterating.
```

Comment:
`fusion->exprs()` creates a new list: https://github.com/NVIDIA/Fuser/blob/9780f666e55ef6a2d76b8da1396a2227c8c37e69/csrc/fusion.h#L196. So why was continuing iterating `exprs` a problem?

Reply:
`ir_utils::replaceValInAllExprInputsAndFusionOutputs` deletes and creates new expressions, so `fusion->exprs()` list holds pointers to deleted expression. We can work around it, by updating the expression pointers as we mutate them, but it would be nice if we could simply mutate the expression